resources/type-system/inference.md

```diff
@@ -1046,22 +1046,27 @@ The elaboration process sometimes introduces new operations that are not easily
 expressible using the syntax of Dart. To allow these operations to be specified
 succintly, the syntax of Dart is extended to allow the following forms:
 
-- `@AWAIT_WITH_TYPE_CHECK<T>(m_1)`, represents the following operation:
+- `@AWAIT_WITH_TYPE_CHECK(m_1)` represents the following operation:
 
-  - Evaluate `m_1`, and let `v` denote the result.
+  - Let `T_1` be the static type of `m_1`.
+
+  - Let `T` be `flatten(T_1)`.
+
+  - Evaluate `m_1`, and let `v` denote the result. _Note that since `m_1` has
+    static type `T_1`, by soundness, `v` must be an instance satisfying
+    `T_1`. Since `T_1 <: FutureOr<flatten(T_1)>` for all `T_1`, it follows that
+    `v` must be an instance satisfying `FutureOr<flatten(T_1)>`, or
+    equivalently, satisfying `FutureOr<T>`. Therefore, `v` must either be an
+    instance satisfying either `Future<T>` or `T`. We consider the two cases
+    below._
 
   - If `v` is an instance satisfying type `Future<T>`, then
-    `@AWAIT_WITH_TYPE_CHECK<T>(m_1)` evaluates to `v`.
+    `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluates to `v`.
 
   - Otherwise, let `u` be a future satisfying type `Future<T>` that will
     complete to the value `v` at some later point. Then,
-    `@AWAIT_WITH_TYPE_CHECK<T>(m_1)` evaluates to `u`.
-
-    - _TODO(paulberry): explain why such a future is guaranteed to (soundly)
-      exist, by stipulating that if `T_1` is the static type of `m_1`, then `T`
-      must be `flatten(T_1)`, and then proving a lemma that `T_1 <:
-      FutureOr<flatten(T_1)>`; therefore `T_1 <: FutureOr<T>`, so in this
-      "otherwise" case, `v` must be an instance satisfying `T`._
+    `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluates to `u`. _Such a future is guaranteed
```

Comment:
Hmm, I think we may have had a miscommunication here. I'm not trying to claim that the reader can pick a `u` satisfying certain properties, and no matter which one they pick, `@AWAIT_WITH_TYPE_CHECK(m_1)` will evaluate to it.

Rather, I'm trying to _define_ the runtime behavior of `@AWAIT_WITH_TYPE_CHECK(m_1)`, but I'm trying to do it in a way that allows for some freedom in the implementation. What is required in order for the implementation to be conformant to the spec is for `@AWAIT_WITH_TYPE_CHECK(m_1)` to evaluate to a `v` that has these certain properties.

Does that make sense?

Reply:
Ok, I've spent some more time thinking about this idea, and I think the problem I'm hung up on is that I'm trying to describe the operation of `@AWAIT_WITH_TYPE_CHECK` procedurally. I'm saying "do this, then do that, then do that." If I were defining `@AWAIT_WITH_TYPE_CHECK` as a separate function all its own, then the last step in the function would be "then return `u`", and I think it would be clear to everyone that what I mean by "then return `u`" is "then the implementation should continue with its evaluation of whatever expression or statement contained the invocation of `@AWAIT_WITH_TYPE_CHECK`, considering `u` to be the result of that invocation". But I don't want to say "then return `u`" because I'm _not_ defining `@AWAIT_WITH_TYPE_CHECK` as a separate function, so I'm worried that a reader might misinterpret "then return `u`" as meaning "now exit whatever function is currently being compiled, with a return value of `u`", and that's definitely _not_ what I mean.

So the wording I went with was "`@AWAIT_WITH_TYPE_CHECK(m_1)` evaluates to `u`". Since the context of this text is "`@AWAIT_WITH_TYPE_CHECK(m_1)` represents the following operation:", I thought it would be clear enough that what I mean here is "and now we're done with the operation I'm defining, and `u` is the resulting value." I really didn't expect anyone to think I was suggesting that "for all `u` satisfying certain properties, `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluates to that `u`".

I definitely don't think my phrasing is great, and I'm open to discussing alternatives. But I'm not understanding how to fit your suggestion into a procedural description of the operation.

Are you suggesting that I start off with, "Let `u` be the result of evaluating `@AWAIT_WITH_TYPE_CHECK(m_1)`. Then `u` has the following properties...", like this?

> - `@AWAIT_WITH_TYPE_CHECK(m_1)` represents the following operation. Let `u` be the result of evaluating `@AWAIT_WITH_TYPE_CHECK(m_1)`. Then `u` has the following properties:
>
>   - Let `T_1` be the static type of `m_1`.
>
>   - Let `T` be `flatten(T_1)`.
>
>   - Evaluate `m_1`, and let `v` denote the result. _Note that since `m_1` has static type `T_1`, by soundness, `v` must be an instance satisfying `T_1`. Since `T_1 <: FutureOr<flatten(T_1)>` for all `T_1`, it follows that `v` must be an instance satisfying `FutureOr<flatten(T_1)>`, or equivalently, satisfying `FutureOr<T>`. Therefore, `v` must either be an instance satisfying either `Future<T>` or `T`. We consider the two cases below._
>
>   - If `v` is an instance satisfying type `Future<T>`, then `u` is equal to `v`.
>
>   - Otherwise, `u` is a future satisfying type `Future<T>` that will complete to the value `v` at some later point. _Such a future is guaranteed to (soundly) exist, because in this case `v` is an instance satisfying `T`._

This approach seems really awkward to me, because it no longer feels like a procedural description of an operation, and IMHO a procedural description of an operation is really going to be the most straightforward thing to do here.

CC @lrhn @leafpetersen @eernstg in case any of you have other ideas for how to phrase this more clearly.

Reply:
Ah, now I understand what you're getting at. It wasn't clear to me that this was an implementation procedure. In that case, I might say something like "*let* `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluate to `u`." We can also just omit the name:

"Otherwise, `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluates to a future satisfying type `Future<T>` that will complete to the value `v` at some later point."

I think now my biggest issue (and part of the reason I was confused before) is that we merely claim existence of such a future, but an algorithm should probably offer something more constructive? I know we're trying to avoid overspecifying implementation details, so we can't just say it evaluates `Future.value(v)`. But perhaps we should at least mention that as a witness to the existence property?

Reply:
That makes sense, thanks! I've changed it to:

>   - If `v` is an instance satisfying type `Future<T>`, then let `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluate to `v`.
>
>   - Otherwise, let `@AWAIT_WITH_TYPE_CHECK(m_1)` evaluate to a future satisfying type `Future<T>` that will complete to the value `v` at some later point. _Such a future could, for instance, be created by executing `Future<T>.value(v)` (this is sound because in this case `v` is an instance satisfying `T`). It also could be created using a more efficient implementation-specific mechanism._